sdk/clientcore/core/src/main/java/io/clientcore/core/http/pipeline/OAuthBearerTokenAuthenticationPolicy.java

```diff
@@ -0,0 +1,124 @@
+    }
+
+    /**
+     * Authorizes the request with the bearer token acquired using the specified {@code tokenRequestContext}
+     *
+     * @param request the HTTP request.
+     * @param tokenRequestContext the token request context to be used for token acquisition.
+     */
+    protected void setAuthorizationHeader(HttpRequest request, OAuthTokenRequestOptions tokenRequestContext) {
+        setAuthorizationHeaderHelper(request, tokenRequestContext, true);
+    }
+
+    private void setAuthorizationHeaderHelper(HttpRequest httpRequest, OAuthTokenRequestOptions tokenRequestContext,
+        boolean checkToForceFetchToken) {
+        AccessToken token = credential.getToken(tokenRequestContext);
+        setAuthorizationHeader(httpRequest.getHeaders(), token.getToken());
+    }
+
+    private static void setAuthorizationHeader(HttpHeaders headers, String token) {
+        headers.set(HttpHeaderName.AUTHORIZATION, BEARER + " " + token);
+    }
+
+    @Override
+    public Response<?> process(HttpRequest httpRequest, HttpPipelineNextPolicy next) {
+        if (!"https".equals(httpRequest.getUri().getScheme())) {
+            throw LOGGER.logThrowableAsError(
+                new RuntimeException("Token credentials require a URL using the HTTPS protocol scheme"));
+        }
+
+        HttpPipelineNextPolicy nextPolicy = next.copy();
+
+        authorizeRequest(httpRequest);
+        Response<?> httpResponse = next.process();
+        String authHeader = httpResponse.getHeaders().getValue(HttpHeaderName.WWW_AUTHENTICATE);
+        if (httpResponse.getStatusCode() == 401 && authHeader != null) {
+            if (authorizeRequestOnChallenge(httpRequest, httpResponse)) {
+                // body needs to be closed or read to the end to release the connection
+                try {
+                    httpResponse.close();
+                } catch (IOException e) {
+                    throw LOGGER.logThrowableAsError(new RuntimeException(e));
+                }
+                return nextPolicy.process();
+            } else {
+                return httpResponse;
+            }
+        }
+        return httpResponse;
+    }
+
+    /**
+     * Handles the authentication challenge in the event a 401 response with a WWW-Authenticate authentication challenge
+     * header is received after the initial request and returns appropriate {@link OAuthTokenRequestOptions} to be
+     * used for re-authentication.
+     *
+     * <p>
+     * The default implementation doesn't handle challenges. You can override and your implementation as needed.
+     * </p>
+     *
+     * @param httpRequest The http request.
+     * @param response The Http Response containing the authentication challenge header.
+     * @return A boolean indicating if containing the {@link OAuthTokenRequestOptions} for re-authentication
+     */
+    public boolean authorizeRequestOnChallenge(HttpRequest httpRequest, Response<?> response) {
```

Comment:
Are we missing tests for these classes in clientcore?

Reply:
The use cases for this policy are still to be determined, tests will be added in follow up PR.